### PR TITLE
TLS integration test

### PR DESCRIPTION
**Pull request by marceloneppel**
# Issue
* Jira issue: [DPE-511](https://warthogs.atlassian.net/browse/DPE-511)
* PostgreSQL charm needs to have a way to enable TLS in the connections between instances and from the client applications to the PostgreSQL server.
* To ensure that implementation works, an integration test should exist.

# Solution
* This PR finished to solve that issue by implementing the integrations tests for the TLS implementation made in the previous PRs.

# Context
*This PR consists on `REPLICATING` what was already reviewed on https://github.com/canonical/postgresql-k8s-operator/pull/37 to the PostgreSQL VM charm.
* You can `FOCUS` your review on:
  * `tests/integration/helpers.py`
  * `tests/integration/test_tls.py`
  * Both above files have pretty similar changes to the ones made on https://github.com/canonical/postgresql-k8s-operator/pull/37 (except the test with the Mattermost charm, that doesn't exist on this charm).
  * The difference between this and the k8s charm PR is described below:
    * I added one more check on https://github.com/canonical/postgresql-operator/pull/31/commits/5736ea2b0bcd3020d2c480aa9a61c908d2801049, focusing on [pg_stat_ssl view](https://www.postgresql.org/docs/12/monitoring-stats.html) that tracks the connections that are being made to PostgreSQL (and whether they are encrypted).

# Testing
* Integration test for TLS implementation was added. It tests the connection using TLS through psycopg2.

# Release Notes
* Add TLS implementation integration test.

### CODE REVIEW

File: tests/integration/test_tls.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        # Relate it to the PostgreSQL to enable TLS.
+        await ops_test.model.relate(DATABASE_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+        await ops_test.model.wait_for_idle(status="active", timeout=1000)
+
+        # Wait for all units enabling TLS.
+        for unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+            assert await check_tls(ops_test, unit.name, enabled=True)
+
+        # Remove the relation.
```
Comment (WRFitch) [inline tests/integration/test_tls.py:47]: Can you send some commands to guarantee traffic is encrypted?

**Review by WRFitch — APPROVED** [review]
LGTM, but we could be more thorough about verifying transmission is encrypted.

File: tests/integration/test_tls.py (hunk 1 above)
Comment (marceloneppel) [reply tests/integration/test_tls.py:47]: Good suggestion. I added one more check on https://github.com/canonical/postgresql-operator/pull/31/commits/5736ea2b0bcd3020d2c480aa9a61c908d2801049, focusing on [pg_stat_ssl view](https://www.postgresql.org/docs/12/monitoring-stats.html) that tracks the connections that are being made to PostgreSQL (and whether they are encrypted).

**Review by WRFitch — APPROVED** [review]

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -316,12 +318,15 @@ async def execute_query_on_unit(
...
+        sslmode: Optional ssl mode to use (defaults to None).
 
     Returns:
         A list of rows that were potentially returned from the query.
     """
+    extra_connection_parameters = f" sslmode={sslmode}" if sslmode is not None else ""
     with psycopg2.connect(
-        f"dbname='{database}' user='operator' host='{unit_address}' password='{password}' connect_timeout=10"
+        f"dbname='{database}' user='operator' host='{unit_address}'"
+        f"password='{password}' connect_timeout=10{extra_connection_parameters}"
```
Comment (zmraul) [inline tests/integration/helpers.py:329]: nit: I think is more legible to place the space of the string where you are adding things
```python
extra_connection_parameters = f"sslmode={sslmode}" if sslmode is not None else ""
f"password='{password}' connect_timeout=10 {extra_connection_parameters}"
```

**Review by zmraul — APPROVED** [review]
Looks good. A small nit :) Also not necessarily linked to this PR, but I think we need to add more logging on the integration tests as well.

File: tests/integration/helpers.py (hunk 2 above)
Comment (marceloneppel) [reply tests/integration/helpers.py]: Yeah, I agree that we need more logging on our tests. I will improve on the next PRs. Thanks for the suggestions. I updated the code on https://github.com/canonical/postgresql-operator/pull/31/commits/83b95e8d9fb4506b62332761365afaecc087a20b.

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -316,12 +318,15 @@ async def execute_query_on_unit(
         password: The PostgreSQL superuser password.
         query: Query to execute.
         database: Optional database to connect to (defaults to postgres database).
+        sslmode: Optional ssl mode to use (defaults to None).
 
     Returns:
         A list of rows that were potentially returned from the query.
     """
+    extra_connection_parameters = f" sslmode={sslmode}" if sslmode is not None else ""
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply tests/integration/helpers.py]: Of course you can copy it or part of it if it's useful for MySQL :). Thanks for the review Paulo. I fixed that on https://github.com/canonical/postgresql-operator/pull/31/commits/83b95e8d9fb4506b62332761365afaecc087a20b.

**Pull request merged**